core/drivers/crypto/hisilicon/hisi_qm.c

```diff
@@ -428,11 +437,67 @@ static enum hisi_drv_status qm_alloc(struct hisi_qm *qm)
 	memset(qm->cqc, 0, cqc_size);
 	qm->cqc_dma = virt_to_phys(qm->cqc);
 
-	qm->qp_array = calloc(qm->qp_num, sizeof(struct hisi_qp));
+	return HISI_QM_DRVCRYPT_NO_ERR;
+
+	free(qm->cqc);
+free_sqc:
+	free(qm->sqc);
+	return ret;
+}
+
+static enum hisi_drv_status hisi_qm_alloc_rsv_xqc(struct hisi_qm *qm)
+{
+	enum hisi_drv_status ret = HISI_QM_DRVCRYPT_NO_ERR;
+
+	qm->rsv_xqc.sqc = (struct qm_sqc *)memalign(HISI_QM_ALIGN32,
+						    sizeof(struct qm_sqc));
+	if (!qm->rsv_xqc.sqc) {
+		EMSG("Fail to malloc rsv sqc!");
+		return HISI_QM_DRVCRYPT_ENOMEM;
+	}
+	memset(qm->rsv_xqc.sqc, 0, sizeof(struct qm_sqc));
+	qm->rsv_xqc.sqc_dma = virt_to_phys(qm->rsv_xqc.sqc);
+
+	qm->rsv_xqc.cqc = (struct qm_cqc *)memalign(HISI_QM_ALIGN32,
+						    sizeof(struct qm_cqc));
+	if (!qm->rsv_xqc.cqc) {
+		EMSG("Fail to malloc rsv cqc!");
+		ret = HISI_QM_DRVCRYPT_ENOMEM;
+		goto free_rsv_sqc;
+	}
+	memset(qm->rsv_xqc.cqc, 0, sizeof(struct qm_cqc));
+	qm->rsv_xqc.cqc_dma = virt_to_phys(qm->rsv_xqc.cqc);
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+
+	free(qm->rsv_xqc.cqc);
+free_rsv_sqc:
+	free(qm->rsv_xqc.sqc);
+	return ret;
+}
+
+static enum hisi_drv_status qm_alloc(struct hisi_qm *qm)
+{
+	enum hisi_drv_status ret = HISI_QM_DRVCRYPT_NO_ERR;
+	int32_t j;
+	size_t qp_size;
+	uint32_t i;
+
+	qp_size = sizeof(struct hisi_qp) * qm->qp_num;
+
+	ret = hisi_qm_alloc_xqc(qm);
+	if (ret)
+		return ret;
+
+	ret = hisi_qm_alloc_rsv_xqc(qm);
+	if (ret)
+		goto free_xqc;
+
+	qm->qp_array = calloc(qm->qp_num, qp_size);
```

Comment:
`qp_size` already contains one multiplication with `qm>qp_num`, do you need a second one?

Reply:
do not need.solved.